### PR TITLE
Feature read json

### PR DESCRIPTION
**Pull request by ltalirz**

### CODE REVIEW

File: digitizer/config.py [hunk 1]
```diff
@@ -55,6 +55,18 @@ def find_by_name(name, json):
 
     raise ValueError('JSON for {} not found.'.format(name))
 
+
+def find_by_key(keyvalue, keytype, json):
```
Comment (ltalirz) [inline digitizer/config.py]:
```suggestion
def find_by_key(value, key, json):
```

File: digitizer/forms.py [hunk 2]
```diff
@@ -9,6 +9,7 @@
 from .config import QUANTITIES
 from .adsorbates import Adsorbates
 from .parse import prepare_isotherm_dict
+from .read_isotherm import read_isotherm_json
```
Comment (ltalirz) [inline digitizer/forms.py]:
```suggestion
from .read_json import read_isotherm_json
```

File: digitizer/forms.py [hunk 3]
```diff
@@ -71,6 +72,9 @@ def __init__(self, plot, tabs):  # pylint: disable=redefined-outer-name
         self.inp_digitizer = pw.TextInput(name='Digitizer',
                                           placeholder='Your full name')
 
+        # fill form from JSON upload
+        self.inp_jsondata = pw.FileInput(name='Upload JSON Isotherm')
```
Comment (ltalirz) [inline digitizer/forms.py]:
```suggestion
        self.inp_json = pw.FileInput(name='Upload JSON Isotherm')
```

File: digitizer/forms.py [hunk 4]
```diff
@@ -81,35 +85,31 @@ def __init__(self, plot, tabs):  # pylint: disable=redefined-outer-name
         self.btn_plot = pn.widgets.Button(name='Plot', button_type='primary')
         self.btn_plot.on_click(self.on_click_plot)
 
+        self.btn_xferjson = pn.widgets.Button(name='Transfer data from JSON',
```
Comment (ltalirz) [inline digitizer/forms.py:88]: Another option is to use the "prefill" button for this.
Put the JSON upload next to his button. The prefill callback will check whether a file has been uploaded. If not, it uses sample data (by the way, ideally "prefill" would parse a sample JSON, which would then also become a test in CI).

File: digitizer/read_isotherm.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                             QUANTITIES[species_type]['json'])
+    elif 'name' in species.keys():
+        # Fall back on name
+        output = species
+    else:
+        raise ValueError('Adsorbate or Adsorbent metadata in JSON incomplete')
+    return output['name']
+
+
+def read_isotherm_json(form):  # pylint: disable=unused-argument, disable=too-many-branches, disable=too-many-statements
```
Comment (ltalirz) [inline digitizer/read_isotherm.py]:
```suggestion
def load_isotherm_json(form):  # pylint: disable=unused-argument, disable=too-many-branches, disable=too-many-statements
```

File: digitizer/read_isotherm.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    # At present, only for single-component isotherms
```
Comment (ltalirz) [inline digitizer/read_isotherm.py]: Is this still valid? I see some multicomponent code below...

**Review by ltalirz — COMMENTED** [review]
thanks @dwsideriusNIST !

I took the liberty to open a PR from your branch so that I could comment on it.

Besides some simplification in the form, I think the main thing is to add two example JSONs to the test folder (one single-component and one multi-component) and to use them as part of the CI to test the `read_isotherm_json` function

File: digitizer/read_isotherm.py (hunk 6 above)
Comment (dwsideriusNIST) [reply digitizer/read_isotherm.py]: no, I neglected to remove it... fixed momentarily

Comment (dwsideriusNIST) [reply digitizer/read_isotherm.py]: done

**Comment by dwsideriusNIST** [comment]
I folded all the requested changes into the feature-read_json branch. I will address the question of which button to use for loading the input JSON (noted above). The other issue is that an error message was being shown in the "input json" area of the form.

File: digitizer/forms.py (hunk 4 above)
Comment (dwsideriusNIST) [reply digitizer/forms.py:88]: excellent idea. I have a handful of test JSON files that can be co-opted for the tests:
1. basic single component
2. single component in log-scale pressure
3. full quaternary multicomponent (all compositions specified, isotherms for all species)
4. "partial" quaternary multicomponent (four species specified, but isotherm for only one species included [only has composition for that species]) - odd case, but we find isotherms like this.

**Comment by ltalirz** [comment]
Thanks @dwsideriusNIST , I'm happy to take it from here.
Can you link somewhere the sample JSON files you were talking about (or perhaps make a separate PR to add them to the `tests/static` folder?)

**Comment by ltalirz** [comment]
Thanks, Dan. I will push your branch directly to the repository and keep working from there.

**Comment by ltalirz** [comment]
Superseded by https://github.com/ltalirz/isotherm-digitizer-panel/pull/52

**Pull request closed without merging**